### PR TITLE
Function calls for fallback paths

### PR DESCRIPTION
**Pull request by Krovatkin**
This PR adds API to package unoptimized/fallback blocks as function calls. It's mainly meant to be used by TensorExpressionsFuser and SpecializeAutogradZero passes as both specialize the original graph but would also like to provide a fallback path in case the assumptions under which the graph was specialized do not hold for some inputs.

### CODE REVIEW

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 1]
```diff
@@ -402,6 +404,9 @@ ExecutionPlan ProfilingGraphExecutorImpl::getPlanFor(
 
   auto copy = pr_->graph()->copy();
   runProfilingOptimizations(copy);
+  // replaces a fallback graph inserted by
+  // if specialize_autogradzero one exists
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:408]: nit: replaces a fallback graph inserted by specialize_autogradzero if one exists

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 2]
```diff
@@ -416,5 +421,85 @@ GraphExecutorState ProfilingGraphExecutorImpl::getDebugState() {
   return state;
 }
 
+Node* createFallbackGraph(Block* b, ArrayRef<Value*> inputs, Graph* g) {
+  auto graph = std::make_shared<Graph>();
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:425]: We might want to assert that `!b->owningNode() || b->owningNode()->kind() == prim::If` - it would invalid to run this on a loop block

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 3]
```diff
@@ -416,5 +421,85 @@ GraphExecutorState ProfilingGraphExecutorImpl::getDebugState() {
...
+
+void ProfilingGraphExecutorImpl::replaceFallbackGraphWithFallbackFunction(
+    Block* b) {
+  Stack s;
+  for (auto it = b->nodes().begin(); it != b->nodes().end();) {
+    if (it->kind() == prim::FallbackGraph) {
+      auto fallback_func = createFallbackPathFunction(
+          it->g(attr::Subgraph)->block(), "fallback_function");
+      fallback_func->get_executor().getPlanFor(s, bailout_depth - 1);
+      fallback_functions_.emplace_back(fallback_func);
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:487]: fallback_func_ are responsible for owning this memory, correct ?

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.h [hunk 4]
```diff
@@ -4,6 +4,11 @@
 namespace torch {
 namespace jit {
 
+TORCH_API Node* createFallbackGraph(
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.h:7]: This is an unused API atm - do you mind adding a test when you plan to land this

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 5]
```diff
@@ -416,5 +421,85 @@ GraphExecutorState ProfilingGraphExecutorImpl::getDebugState() {
...
+  auto value_map = [](Value* v) { return v; };
+  auto graph = std::make_shared<Graph>();
+  graph->block()->cloneFrom(b, value_map);
+
+  auto otypes = c10::fmap(
+      graph->return_node()->inputs(), [](Value* v) { return v->type(); });
+  auto tuple_type = TupleType::create(otypes);
+  auto return_tuple = graph->createTuple(graph->return_node()->inputs());
+  graph->appendNode(return_tuple);
+  for (int i = graph->outputs().size() - 1; i >= 0; i--) {
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:451]: I take it we need to erase these outputs bc functions can only have one output ? maybe add one line -comment

File: torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp [hunk 6]
```diff
@@ -45,6 +45,15 @@ RegisterOperators reg(
            };
          },
          aliasAnalysisSpecialCase()),
+     Operator(
+         prim::FallbackGraph,
+         [](const Node* node) -> Operation {
+           return [](Stack* stack) {
+             AT_ERROR(
+                 "Must be lowered to Interpreter's PROFILE instruction"); // NOLINT
```
Comment (eellison) [inline torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:53]: bad copy-pasta

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.h [hunk 7]
```diff
@@ -17,10 +22,12 @@ struct ProfilingGraphExecutorImpl : public GraphExecutorImplBase {
  private:
   void runProfilingInsensitiveOptimizations(std::shared_ptr<Graph>& graph);
   void runProfilingOptimizations(std::shared_ptr<Graph>& graph);
+  void replaceFallbackGraphWithFallbackFunction(Block* b);
   std::unique_ptr<ProfilingRecord> pr_;
   c10::optional<ExecutionPlan>
       profiling_plan_; // plan to run in order to profiling the code
   c10::optional<ExecutionPlan> optimized_plan_;
+  std::vector<std::unique_ptr<Function>> fallback_functions_;
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.h:30]: nit: maybe add a comment that these functions are owned by this interpreter, although i guess the unique_ptr makes that evident

**Review by eellison — COMMENTED** [review]
Looks great !! Very clean and simple.

Would it be worth considering moving the new functions to a separate file ? 

Also, can you add some comments somewhere about why we're doing prim::FallbackFunction instead of a function call ? (so we dont wreck alias analysis / more transparent IR)

We still need to implement the interpreter changes, and add tests i think

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp (hunk 2 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:425]: a good point!

File: torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp (hunk 6 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:53]: ouch

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.h (hunk 7 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.h:30]: will add a comment

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp (hunk 3 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:487]: yes, it's very similar to how Interpreter does it and seems pretty straightforward.

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp (hunk 5 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:451]: yes)))))))) bailouts followed the exact same pattern. will add a comment.

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp (hunk 1 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:408]: will fix

**Review by eellison — COMMENTED** [review]
One thing that just occurred to me is we can only analyze FallBackGraphs as we are right now bc at the point of optimizing a graph `A` that contains a FallbackGraph `B`, we know we haven't yet optimized `B` and thus `B` doesnt contain any FallbackGraphs. If we were to start to reprofile or something and `B` had been optimized at the point of us analyzing `A` than we would suddenly be in the business of analyzing all of our specializations graphs at once recursively 🥵 

I think we're fine for now but something to keep in mind. If we get to such a situation I think we could probably make outputs of the graph alias inputs and assume no writes, bc atm we're only creating this 
in write-free contexts (Backward Graph, Fusion Groups)

File: test/cpp/jit/test_misc.cpp [hunk 8]
```diff
@@ -1196,6 +1198,76 @@ void testThreadLocalDebugInfo() {
...
+static void nestGraphIntoFallbackGraph(std::shared_ptr<Graph> graph) {
+  auto fallback =
+      createFallbackGraph(graph->block(), graph->inputs(), graph.get());
+  graph->prependNode(fallback);
+  for (size_t i = 0; i < graph->outputs().size(); i++) {
+    graph->outputs()[i]->replaceAllUsesWith(fallback->output(i));
+    fallback->output(i)->copyMetadata(graph->outputs()[i]);
+  }
+  GRAPH_DUMP("nestGraphIntoFallbackGraph:", graph);
+  EliminateDeadCode(graph);
```
Comment (eellison) [inline test/cpp/jit/test_misc.cpp:1210]: I don't think all callers of the API should have to call EliminateDeadCode after calling. It's going to make working on these passes a little more confusing if there is a bunch of dead code left around after every FallBack path is created.

File: test/cpp/jit/test_misc.cpp [hunk 9]
```diff
@@ -1196,6 +1198,76 @@ void testThreadLocalDebugInfo() {
...
+      }
+
+      f.run(stack);
+    }
+  }
+
+  at::Tensor at;
+  pop(stack, at);
+  float af = et.item<float>();
+  ASSERT_EQ(af, ef);
```
Comment (eellison) [inline test/cpp/jit/test_misc.cpp:1268]: Why is the value expected to be the same if we're initializing random variables and running the graph with them ?

File: test/cpp/jit/test_misc.cpp [hunk 10]
```diff
@@ -1196,6 +1198,76 @@ void testThreadLocalDebugInfo() {
...
+        auto fallback = *it++;
+        ASSERT_EQ(it, opt_graph->block()->nodes().end());
+        ASSERT_TRUE(fallback->hasAttribute(attr::Subgraph));
+        testing::FileCheck()
+            .check("Tensor = aten::mul")
+            ->check("Tensor = aten::mul")
+            ->run(*fallback->g(attr::Subgraph));
+      }
+
+      f.run(stack);
```
Comment (eellison) [inline test/cpp/jit/test_misc.cpp:1261]: We don't pop the value off the stack in the first runs

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 11]
```diff
@@ -416,5 +421,87 @@ GraphExecutorState ProfilingGraphExecutorImpl::getDebugState() {
...
 
+Node* createFallbackGraph(Block* b, ArrayRef<Value*> inputs, Graph* g) {
+  auto graph = std::make_shared<Graph>();
+  auto value_map = [](Value* v) { return v; };
+  graph->block()->cloneFrom(b, value_map);
+
+  auto fallback = g->create(prim::FallbackGraph, inputs, b->outputs().size());
+  fallback->g_(attr::Subgraph, graph);
+
+  for (size_t i = 0; i < b->outputs().size(); i++) {
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:432]: Can you please clean up the graph here and remove the old block? you can just iterate from the end of the block -> beginning and destroy every node

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 12]
```diff
@@ -416,5 +421,87 @@ GraphExecutorState ProfilingGraphExecutorImpl::getDebugState() {
   return state;
 }
 
+Node* createFallbackGraph(Block* b, ArrayRef<Value*> inputs, Graph* g) {
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:424]: Can you change the API to `replaceBlockWithFallBackGraph` ? I don't think there's ever going to be an API callsite where we don't want to do this

**Review by eellison — APPROVED** [review]
LGTM - it was a little bit hard to validate that the fallback graphs are being run bc i'm not sure exactly how modifying last executed graph runs, but anyway we'll have further validation & it will be easier to test when we merge this. 

One request is to change the API a little bit, and delete the dead code as part of the function call creation. I think we'll be doing ourselves a favor if we add small amounts of work to remove dead information in the graph as we're working on it. When I was first working on specialize autogradzero I was confused for a while by the existing unoptimized block of code that was left at the end of the graph but is unused...

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:425]: actually, dont think this invalid to run on a loop block it would just be weird lol. so no assertion is fine...

**Comment by eellison** [comment]
also, a PR description would be nice

File: test/cpp/jit/test_misc.cpp (hunk 8 above)
Comment (Krovatkin) [reply test/cpp/jit/test_misc.cpp:1211]: this is a helper function for this particular test; I was actually hoping that it expresses the intent better than manually removing nodes. Fixed it

File: test/cpp/jit/test_misc.cpp (hunk 10 above)
Comment (Krovatkin) [reply test/cpp/jit/test_misc.cpp:1257]: fixed

File: test/cpp/jit/test_misc.cpp (hunk 9 above)
Comment (Krovatkin) [reply test/cpp/jit/test_misc.cpp:1269]: errr, it's fixed now

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp (hunk 2 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:456]: yup it's perfectly valid; sorry i should've posted earlier when i revisited your suggestion

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp (hunk 11 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:463]: errr, that doesn't seem like a valid thing to do (not always, at least)
here's how we call this API from `tensorexpr_fuser.cpp`

```cpp
    WithInsertPoint guard(false_block->return_node());
    auto fallback = createFallbackGraph(subgraph->block(), n->inputs(), n->owningGraph());
    n->owningGraph()->insertNode(fallback);
```

where `subgraph` comes from 

```
  void guardFusionGroup(Node* n) {
    auto subgraph = SubgraphUtils::getSubgraph(n);
```

which we shouldn't be mocking with.

Frankly, I think it should be a responsibility of a caller to clean up a graph if required. `createFallbackGraph` (given its name) should just create a fallback graph. We can provide an extra API to clean up a graph.

Comment (eellison) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:469]: You could just as easily write the same three lines by copying the fallback graph into the block, and then calling `replaceBlockWithFallBackGraph`. I have run into a few places where we leave the graph in a bad state and it's taking me longer to work on the code base as a result. We have 5 duplicated profiles for every node, specialize_autogradzero had the whole graph copied and not cleaned up, I think we should prioritize is this. This is very easy logic to add but makes it easier on pass writers and on debugging the graph. 

Additionally, calling `EliminateDeadCode` is not strictly safe when you're duplicating the block. What if the block has mutation, or a print statement ? These nodes will not get removed by DCE. We need to remove the nodes as part of the API.

**Review by eellison — CHANGES_REQUESTED** [review]
Realized current API isn't safe if there's side-effectful nodes in the fallbackpath

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 13]
```diff
@@ -65,6 +65,21 @@ std::atomic<size_t>& getBailoutDepth() {
   return bailout_depth;
 }
 
+void removeProfilingNodes(Block* b) {
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:68]: This is an unrelated change, why is it in here, could you remove.

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 14]
```diff
@@ -382,12 +399,16 @@ ExecutionPlan ProfilingGraphExecutorImpl::getPlanFor(
   std::lock_guard<std::mutex> lock(compile_mutex);
   GRAPH_DEBUG("Running ProfilingGraphExecutorImpl ", this);
 
+  if (!remaining_bailout_depth_.has_value() || !tensorExprFuserEnabled()) {
+    remaining_bailout_depth_ = remaining_bailout_depth;
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:403]: Could you add a comment here ? Why is this here ?

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 15]
```diff
@@ -417,9 +438,13 @@ ExecutionPlan ProfilingGraphExecutorImpl::getPlanFor(
   auto copy = pr_->graph()->copy();
   ProfilingRecord::removeProfileCounter(copy->block());
   runProfilingOptimizations(copy);
+  // replaces a fallback graph inserted by
+  // specialize_autogradzero if one exists
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:442]: `specialize_autogradzero if one exists` outdated call ?

**Review by eellison — APPROVED** [review]
LGTM - there are some unrelated change here, could you comment on why they're added / potentially remove ? may not be needed on master.

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp (hunk 13 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:68]: I'll remove it. it's somewhat related, but it will be only needed when we start using fallbacks in `specialize_autogradzero.cpp`

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp (hunk 14 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:403]: yeah let me add a comment
